### PR TITLE
8276066: Reset LoopPercentProfileLimit for x86 due to suboptimal performance

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

I'd like to reset the value of `LoopPercentProfileLimit` (from 30 to the original 10) for x86 to fix performance degradation.

We had observed that for the same Java App, the performance of x86 is slower than that of aarch64.
But the x86's performance should not be so worse than the aarch64 according to some SPEC benchmark results.

After some investigation, it seems that the slowness of x86 is caused by the different default settings of `LoopPercentProfileLimit` (30 for x86, but 10 for other platforms).
If we change `LoopPercentProfileLimit` from 30 to 10, x86 would run faster.

In JDK-8149421, `LoopPercentProfileLimit` [1] was first added and set to be 30 for x86 and 10 for other platforms.
Logically, the default value of `LoopPercentProfileLimit` is 10 for all platforms even before JDK-8149421.
This is because when `LoopPercentProfileLimit=10`, `10.0` [2] equals `100.0 / LoopPercentProfileLimit` [3].
So if we set `LoopPercentProfileLimit=10`, this unrolling rule [3] would be the same as the original design before JDK-8149421.

One most important fact is that from the very beginning of OpenJDK source code, the default value of `LoopPercentProfileLimit` (logically) is 10 for all platforms.   
So I suggest resetting `LoopPercentProfileLimit` to the original value (10) for x86, just as other platforms.

I've noted that the review thread mentioned that JDK-8149421 would be beneficial for some SPECjvm2008 benchmarks [4].
Then I run SPECjvm2008 with `LoopPercentProfileLimit=10` finding that there is no performance drop on x86.
So it won't revert JDK-8149421's opts for SPECjvm2008.

To show the potential improvement of this change, I've made a jmh test in the patch.
Performance can be improved by 1.25x ~ 2.0x according to this micro benchmark.

Any comments?

Thanks.
Best regards,
Jie

  
[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/loopTransform.cpp#L908
[2] https://github.com/openjdk/jdk8u/blob/master/hotspot/src/share/vm/opto/loopTransform.cpp#L673
[3] https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/loopTransform.cpp#L903
[4] https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2016-February/021205.html

ratio

before

after

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276066](https://bugs.openjdk.java.net/browse/JDK-8276066): Reset LoopPercentProfileLimit for x86 due to suboptimal performance

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6142/head:pull/6142` \
`$ git checkout pull/6142`

Update a local copy of the PR: \
`$ git checkout pull/6142` \
`$ git pull https://git.openjdk.java.net/jdk pull/6142/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6142`

View PR using the GUI difftool: \
`$ git pr show -t 6142`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6142.diff">https://git.openjdk.java.net/jdk/pull/6142.diff</a>

</details>

### CODE REVIEW

**Comment by TobiHartmann** [comment]
> I'll run this through our performance testing and report back.

Performance results look good.

Is this change still required after re-enabling post loop vectorization?

**Comment by DamonFool** [comment]
> > I'll run this through our performance testing and report back.
> 
> Performance results look good.
> 
> Is this change still required after re-enabling post loop vectorization?



> > I'll run this through our performance testing and report back.
> 
> Performance results look good.
> 
> Is this change still required after re-enabling post loop vectorization?

This is part of loop unrolling rule, so I think it would be better to change it to improve the performance for the current code base.
Then all the future opts on x86 can be evaluated based on that improved version.
And we may backport the change to other repos like jdk11u.
So I would suggest resetting it to 10 on x86.
Thanks.

**Comment by DamonFool** [comment]
Hi all,

I tested our deep learning cluster using jdk17 yesterday.
The time of computing stage can be further reduced by 5% ~ 6% with this patch. 
So it's really worth making this change.
And we also plan to backport it if it is accepted in the jdk mainline.

If post loop vectorization requires higher LoopPercentProfileLimit on x86, we can still re-tune it in that enhancement just like other platforms.
I also think it would be better to improve the base line performance with `LoopPercentProfileLimit=10` when evaluating the performance opts like post loop vectorization on x86.

Thanks.
Best regards,
Jie

**Comment by DamonFool** [comment]
Hi @TobiHartmann ,

The `LoopPercentProfileLimit` was changed from 10 to 30 (logically) on x86 to enable post loop vectorization in JDK-8149421.
However, post loop vectorization was disabled by JDK-8183103 without restoring the original value of `LoopPercentProfileLimit` to 10 on x86.

So do you agree that we'd better restore `LoopPercentProfileLimit=10` if post loop vectorization is disabled?                 

According to the comment by @pfustc  in JDK-8183390, JEP 417: Vector API (Third Incubator) [1] is needed to enable post loop vectorization.
But I'm not sure whether `LoopPercentProfileLimit=30` is still the best choice for x86.
Maybe, it should be re-tuned based on the new implementation too.
So it doesn't matter to restore `LoopPercentProfileLimit=10` in the jdk mainline.

And it seems hard to re-enable post loop vectorization for jdk versions without Vector API (e.g., jdk11).
So for better performance experience on x86, I suggest restoring `LoopPercentProfileLimit=10` if post loop vectorization is hard to be re-enabled, which means we'd better restore `LoopPercentProfileLimit=10` for jdk11 too.

What do you think?
Thanks.

[1] https://github.com/openjdk/jdk/pull/5873

**Review by TobiHartmann — APPROVED** [review]
> So do you agree that we'd better restore LoopPercentProfileLimit=10 if post loop vectorization is disabled?

Yes, that seems reasonable to me. We should re-evaluate the default value once post loop vectorization is fixed.

File: test/micro/org/openjdk/bench/vm/compiler/LoopUnroll.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.concurrent.TimeUnit;
+
+@Warmup(iterations = 3, time = 5, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 4, time = 5, timeUnit = TimeUnit.SECONDS)
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@State(Scope.Thread)
+@Fork(value=1)
+public class LoopUnroll {
+  @Param({"16", "32", "64", "128", "256", "512", "1024"})
```
Comment (TobiHartmann) [inline test/micro/org/openjdk/bench/vm/compiler/LoopUnroll.java:37]: We use four whitespace indentation for Java code.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: test/micro/org/openjdk/bench/vm/compiler/LoopUnroll.java (hunk 1 above)
Comment (DamonFool) [reply test/micro/org/openjdk/bench/vm/compiler/LoopUnroll.java]: > We use four whitespace indentation for Java code.

Fixed.
Thanks for your review.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by DamonFool** [comment]
> Looks good to me.

Thanks @TobiHartmann .
Will push it tomorrow if there is no objection.

**Comment by DamonFool** [comment]
> I agree.

Thanks @vnkozlov .
/integrate

**Pull request closed without merging**